### DISCUSSION TITLE
How to output an audio stream defined by liquidsoap to a named pipe

### DISCUSSION THREAD

**Discussion by dycfan** [savonet/liquidsoap#3721] [User Support]
Ask a question:
I'm running liquidsoap on an ubuntu22.04 host. The version number is Liquidsoap 2.1.4.
I need to output the audio stream defined by liquidsoap to the named pipe "/tmp/ysmusic". Tried a lot of functions, including output.ao output.file, but failed. Something like this:

`music_list01 = playlist("/mnt/LocalDisk1/Onlinemusic/草原歌曲",mode="normal",reload_mode="watch")
cymusic_one = crossfade(fade_in=10.,fade_out=10.,music_list01)
cymusic_one = add([switch([({0m0s},Clock)]),cymusic_one])
cymusic_two = fallback([cymusic_one,Error])

output.ao(pipe=“/tmp/ysmusic", cymusic_two)

output.icecast(connection_timeout=10.,timeout=50.,
  icy_metadata="true",
  encoding="CN",
  %ffmpeg(format="mp3", %audio(codec="libmp3lame", b="320k")),
  format="audio/mpeg",
  name="草原高原歌曲精选",
  description="Selection of Prairie Plateau Songs",
  genre="Various",
  host="192.168.9.6",port=8008,password="myhome",
  mount="cymusic",cymusic_two)`
In fact, such code simply doesn't work.

I ask you: What should be the right thing to do?

Thank you for your help. Thank you very much!

**Comment by toots** [comment]
Hi! 

A pipe on unix system is simply a named file. You should be able to use `output.file` to write to it!

**Comment by vitoyucepi** [comment]
Hi @dycfan,
The pipes do indeed work as files.
Consider this example.
1. In the first shell, create and listen to the pipe
   ```sh
   mkfifo /tmp/a;
   cat /tmp/a;
   ```
1. In the second shell, write from the liquidsoap into the pipe
   ```ruby
   output.file(%mp3, '/tmp/a', sine())
   ```

But there's a catch, you can't write to the pipe indefinitely without reading data from it.
If you stop the `cat`, liquidsoap will eventually fail with an error.
```
2024/02/08 04:45:09 [clock.main:2] Source output.file failed while streaming: Lang.Runtime_error { kind: "system", msg: "Broken pipe", pos: [at main.liq, line 1, char 0-35] }!
2024/02/08 04:45:09 [clock.main:2] Raised by primitive operation at Stdlib.output_substring in file "stdlib.ml", line 379, characters 7-40
2024/02/08 04:45:09 [clock.main:2] Called from Pipe_output.chan_output#write_pipe in file "outputs/pipe_output.ml", line 337, characters 8-44
2024/02/08 04:45:09 [clock.main:2] Re-raised at Runtime_error.raise in file "tools/runtime_error.ml", line 48, characters 17-51
2024/02/08 04:45:09 [clock.main:2] Called from Pipe_output.piped_output#send in file "outputs/pipe_output.ml", line 294, characters 11-23
2024/02/08 04:45:09 [clock.main:2] Called from Output.encoded#send_frame.output_chunks.f in file "outputs/output.ml", line 240, characters 10-24
2024/02/08 04:45:09 [clock.main:2] Called from Output.output#output in file "outputs/output.ml", line 174, characters 10-35
2024/02/08 04:45:09 [clock.main:2] Called from Clock.clock#end_tick.(fun) in file "clock.ml", line 300, characters 14-22
2024/02/08 04:45:09 [clock.main:2] 
2024/02/08 04:45:09 [clock:2] Error when leaving output output.file: Lang.Runtime_error { kind: "system", msg: "Broken pipe", pos: [at main.liq, line 1, char 0-35] }!
2024/02/08 04:45:09 [clock.main:3] Streaming loop stopped.
2024/02/08 04:45:09 [main:3] Shutdown started!
```
